src/joystick/hidapi/SDL_hidapi_gip.c

```diff
@@ -0,0 +1,2196 @@
+        ctx->hello_deadline = SDL_GetTicks() + GIP_HELLO_TIMEOUT;
+    }
+
+    device->context = ctx;
+    device->type = SDL_GAMEPAD_TYPE_XBOXONE;
+
+    return true;
+}
+
+static int HIDAPI_DriverGip_GetDevicePlayerIndex(SDL_HIDAPI_Device *device, SDL_JoystickID instance_id)
+{
+    return -1;
+}
+
+static void HIDAPI_DriverGip_SetDevicePlayerIndex(SDL_HIDAPI_Device *device, SDL_JoystickID instance_id, int player_index)
+{
+}
+
+static bool HIDAPI_DriverGip_OpenJoystick(SDL_HIDAPI_Device *device, SDL_Joystick *joystick)
+{
+    GIP_Device *ctx = (GIP_Device *)device->context;
+
+    SDL_AssertJoysticksLocked();
+
+    ctx->left_impulse_level = 0;
+    ctx->right_impulse_level = 0;
+    ctx->left_vibration_level = 0;
+    ctx->right_vibration_level = 0;
+    ctx->rumble_state = GIP_RUMBLE_STATE_IDLE;
+    ctx->rumble_time = 0;
+    ctx->rumble_pending = false;
+    SDL_zeroa(ctx->last_input);
+
+    // Initialize the joystick capabilities
+    joystick->nbuttons = 11;
+    if (ctx->features & GIP_FEATURE_ELITE_BUTTONS) {
+        ctx->paddle_offset = -1;
+        if (ctx->firmware_major_version == 5) {
+            ctx->paddle_offset = 14;
+        } else {
+            ctx->paddle_offset = 28;
+        }
+        if (ctx->paddle_offset >= 14) {
+            ctx->paddle_idx = joystick->nbuttons;
+            joystick->nbuttons += 4;
+        } else {
+            ctx->features &= ~GIP_FEATURE_ELITE_BUTTONS;
+        }
+    }
+    if (ctx->features & GIP_FEATURE_CONSOLE_FUNCTION_MAP) {
+        ctx->share_button_idx = joystick->nbuttons;
+        joystick->nbuttons++;
+    }
+    joystick->naxes = SDL_GAMEPAD_AXIS_COUNT;
+    joystick->nhats = 1;
+
+    return true;
+}
+
+static bool HIDAPI_DriverGip_RumbleJoystick(SDL_HIDAPI_Device *device, SDL_Joystick *joystick, Uint16 low_frequency_rumble, Uint16 high_frequency_rumble)
+{
+    GIP_Device *ctx = (GIP_Device *)device->context;
+
+    // Magnitude is 1..100 so scale the 16-bit input here
+    ctx->left_vibration_level = (Uint8)(low_frequency_rumble / 655);
+    ctx->right_vibration_level = (Uint8)(high_frequency_rumble / 655);
+    ctx->rumble_pending = true;
+
+    return HIDAPI_DriverGip_UpdateRumble(ctx);
+}
+
+static bool HIDAPI_DriverGip_RumbleJoystickTriggers(SDL_HIDAPI_Device *device, SDL_Joystick *joystick, Uint16 left_rumble, Uint16 right_rumble)
+{
+    GIP_Device *ctx = (GIP_Device *)device->context;
+
+    // Magnitude is 1..100 so scale the 16-bit input here
+    ctx->left_impulse_level = (Uint8)(left_rumble / 655);
+    ctx->right_impulse_level = (Uint8)(right_rumble / 655);
+    ctx->rumble_pending = true;
+
+    return HIDAPI_DriverGip_UpdateRumble(ctx);
+}
+
+static Uint32 HIDAPI_DriverGip_GetJoystickCapabilities(SDL_HIDAPI_Device *device, SDL_Joystick *joystick)
+{
+    GIP_Device *ctx = (GIP_Device *)device->context;
+    Uint32 result = 0;
+
+    if (ctx->features & GIP_FEATURE_MOTOR_CONTROL) {
+        result |= SDL_JOYSTICK_CAP_RUMBLE | SDL_JOYSTICK_CAP_TRIGGER_RUMBLE;
```

Comment:
Possibly add a vendor check here for trigger rumble? My experience is that almost all third party vendors that support trigger rumble use a buzzy motor instead of a feel similar to the Microsoft controllers, and some of them, especially the 8BitDo controllers, don't support trigger rumble at all.

Reply:
The tactile nature of the rumble is irrelevant to if we should support it; it's advertised as supported to Windows and Xboxes, and being opinionated about whether or not it should be supported is clearly against the vendors' design.

As for not supporting trigger rumble, this is not part of metadata and would require an extensive quirks table. How bad is it if something says it supports trigger rumble but doesn't?

Reply:
Not too bad. It's not a critical gameplay feature for games and most of the time it's controllers where it wouldn't be expected, like small form factor controllers, etc.